heroku_connect/models.py

```diff
@@ -0,0 +1,274 @@
+    """A QuerySet for trigger log models."""
+
+    def failed(self):
+        """Filter for log records with sync failures."""
+        return self.filter(state=TRIGGER_LOG_STATE['FAILED'])
+
+    def related_to(self, instance):
+        """Filter for all log objects of the same connected model as the given instance."""
+        return self.filter(table_name=instance.table_name, record_id=instance.record_id)
+
+
+TRIGGER_LOG_ACTION = {
+    'INSERT': 'INSERT',
+    'UPDATE': 'UPDATE',
+    'DELETE': 'DELETE',
+}
+"""The type of change that a trigger log object represents."""
+
+
+TRIGGER_LOG_ACTION_CHOICES = sorted((value, value) for value in TRIGGER_LOG_ACTION.values())
+
+
+TRIGGER_LOG_STATE = {
+    'SUCCESS': 'SUCCESS',
+    'MERGED': 'MERGED',
+    'IGNORED': 'IGNORED',
+    'FAILED': 'FAILED',
+    'READONLY': 'READONLY',
+    'NEW': 'NEW',
+    'IGNORE': 'IGNORE',
+    'PENDING': 'PENDING',
+    'REQUEUE': 'REQUEUE',
+    'REQUEUED': 'REQUEUED',
+}
+"""The sync state of the change tracked by a trigger log entry."""
+
+
+TRIGGER_LOG_STATE_CHOICES = sorted((value, value) for value in TRIGGER_LOG_STATE.values())
+
+
+class TriggerLogAbstract(models.Model):
+    """
+    Support for accessing the Heroku Connect Trigger Log data and related actions.
+
+    Heroku Connect uses a Trigger Log table to track local changes to connected models (that is,
+    in the Heroku database. Such changes are recorded as rows in the trigger log and, for
+    read-write mappings, eventually written back to Salesforce.
+
+    Old logs are moved to an archive table (usually after 24 hours), from where they are purged
+    eventually (currently 30 days for paid plans, 7 days for demo). Recent logs are modeled by
+    :class:`TriggerLog`; archived logs by :class:`TriggerLogArchive`.
+
+    The data represented by these models is maintained entirely by Heroku Connect, and is
+    instrumental to its operations; it should therefore not be modified. A possible exception is
+    the ``state`` field, which may be changed as detailed in the `error handling`_ section in the
+    Heroku Connect documentation.
+
+    .. seealso::
+        - :class:`.TriggerLogQuerySet`
+        - `Trigger Log in Heroku Connect docs`_
+
+    .. _Trigger Log in Heroku Connect docs:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#understanding-the-trigger-log
+    .. _error handling:
+        https://devcenter.heroku.com/articles/writing-data-to-salesforce-with-heroku-connect#write-errors
+
+    """
+
+    # I18N / TRANSLATIONS:
+    #
+    # Field names and choices don't use translations, because they refer to (English)technical
+    # terms in Heroku Connect's manual. That connection should be preserved. Trigger Log Models
+    # are for technical folks and not intended to be user-facing.
+
+    # read-only fields
+    id = models.BigIntegerField(primary_key=True, editable=False)
+    created_at = models.DateTimeField(editable=False, null=True)
```

Comment:
None of the fields are editable, are they?
I don't know. I would just drop it and make sure to have them as read_only in the admin.

Reply:
Same as above: it's documentation that the field shouldn't be touched, and a hedge against somebody creating a `ModelForm` and messing things up. I think it's better to make that point right where stuff is defined, not in the admin.